### PR TITLE
Reconnect

### PR DESCRIPTION
**Pull request by SSebo**
Still under construction

- [ ] Reconnect

- [ ] Retry

- [ ] Backoff

### CODE REVIEW

**Comment by SSebo** [comment]
hi, @1c3t3a could you help to figure out why the coverage github action can not pass, but the test github action is ok ?

**Comment by SSebo** [comment]
now coverage github action passed, wired.

**Comment by SSebo** [comment]
hi @1c3t3a, the restart testing case will make other case failed, so I add a new `socket-io-restart.js` to do restart operation. #225 PTAL

**Comment by SSebo** [comment]
@1c3t3a Could you help me to figure out why the test pass but coverage action failed?

File: socketio/src/client/client.rs [hunk 1]
```diff
@@ -468,6 +568,53 @@ mod test {
...
 
+    #[test]
+    fn socket_io_reconnect_integration() -> Result<()> {
+        let url = crate::test::socket_io_restart_server();
+
+        let socket = ClientBuilder::new(url).connect();
+        assert!(socket.is_ok(), "should connect success");
+        let socket = socket.unwrap();
+
+        std::thread::sleep(std::time::Duration::from_millis(200));
```
Comment (nshaaban-cPacket) [inline socketio/src/client/client.rs:579]: nit: why sleep?

File: socketio/src/client/builder.rs [hunk 2]
```diff
@@ -302,24 +284,31 @@ impl ClientBuilder {
...
-            TransportType::Websocket => builder.build_websocket()?,
-            TransportType::WebsocketUpgrade => builder.build_websocket_with_upgrade()?,
-        };
-
-        let inner_socket = InnerSocket::new(engine_client)?;
-
+        let transport_type = self.transport_type.clone();
         let socket = Client::new(
-            inner_socket,
+            Box::new(move || build_socket(transport_type.clone(), builder.clone())),
```
Comment (nshaaban-cPacket) [inline socketio/src/client/builder.rs:289]: (creative vision difference) code smell: crossing concerns, seems very odd to pass a copy of the builder in a function into the underlying client it's trying to build

File: socketio/src/client/client.rs [hunk 3]
```diff
@@ -468,6 +568,53 @@ mod test {
...
+        assert!(r.is_ok(), "should emit restart success");
+
+        let mut is_emit_error = false;
+        for _ in 0..10 {
+            std::thread::sleep(std::time::Duration::from_millis(500));
+            let r = socket.emit("message", json!(""));
+            if r.is_err() {
+                is_emit_error = true; // server is down
+                break;
+            }
```
Comment (nshaaban-cPacket) [inline socketio/src/client/client.rs:594]: (usability question) nit: should the re-connection be completely transparent to end user? As a user if I had to attempt to repeat all of my emits that may make it hard to utilize this feature

File: socketio/src/client/client.rs [hunk 4]
```diff
@@ -468,6 +568,53 @@ mod test {
...
+        }
+
+        assert_eq!(
+            is_emit_error, true,
+            "should emit failed because server is down"
+        );
+
+        for _ in 0..10 {
```
Comment (nshaaban-cPacket) [inline socketio/src/client/client.rs:602]: suggestion: why is this looped? Should the next message right after the failure return success?

File: socketio/src/client/client.rs [hunk 5]
```diff
@@ -585,143 +735,115 @@ mod test {
...
-            packet,
-            Packet::new(
-                PacketId::Event,
-                nsp.clone(),
-                Some("[\"auth\",\"success\"]".to_owned()),
-                None,
-                0,
-                None
-            )
-        );
```
Comment (nshaaban-cPacket) [inline socketio/src/client/client.rs:613]: suggestion: The iterator SHOULD be tested by the unit tests, the message pipe is clever, but more complicated that needed for these tests. 

It's two separate concerns:
- testing behavior (via the iterator)
- callback interface is functioning 

If the callback code breaks, it's much more difficult to debug than the raw iterator implementation.

File: socketio/src/client/client.rs [hunk 6]
```diff
@@ -585,143 +735,115 @@ mod test {
...
-        let packet = packet.unwrap();
-
-        assert_eq!(
-            packet,
-            Packet::new(
-                PacketId::Event,
-                nsp.clone(),
-                Some("[\"Hello from the message event!\"]".to_owned()),
-                None,
-                0,
```
Comment (nshaaban-cPacket) [inline socketio/src/client/client.rs:672]: suggestion: Is using the Iterator interface on Client no longer functional?

**Review by nshaaban-cPacket — COMMENTED** [review]
Like to start off by saying this is a valid implementation, as well as effort went into adding tests, which is much appreciated. 

Concerned about the additional complication (addition of arc/mutex/rwlocks) in the underlying client, as well as increased potential in IllegalActionBeforeOpen edge cases. Neither of which are currently present on master.

Perhaps a wrapper around client would be cleaner separating the auto-reconnect logic from the normal client logic? Auto re-connection seems more of a concern for a wrapper around client rather than complicating the underlying client implementation. 

At the end of the day this is a valid implementation, and for how it's done, it looks good (besides is one usability question). While I may prefer a different approach, decision is up to @1c3t3a . Feel free to merge without my approval.

@SSebo Thanks for the PR, sorry for not documenting purpose of the connect_manual, nor intention of iterators use in the future. I recommenced waiting for @1c3t3a before attempting to implement my comments related to differing creative vision (I've annotated the comments in question), as what you have is valid and functional.

Thank you for properly implementing unit tests for your features and expanding the existing test infrastructure.

File: socketio/src/client/client.rs (hunk 1 above)
Comment (SSebo) [reply socketio/src/client/client.rs:579]: Good catch, I found connect function return not means connect success, the first emit message will not success if not wait.

File: socketio/src/client/client.rs [hunk 7]
```diff
@@ -394,6 +452,47 @@ impl Client {
...
+        std::thread::spawn(move || {
+            // tries to restart a poll cycle whenever a 'normal' error occurs,
+            // it just panics on network errors, in case the poll cycle returned
+            // `Result::Ok`, the server receives a close frame so it's safe to
+            // terminate
+            for packet in self_clone.iter() {
+                if let e @ Err(Error::IncompleteResponseFromEngineIo(_)) = packet {
+                    //TODO: 0.3.X handle errors
+                    self_clone.reconnect();
+                    panic!("{}", e.unwrap_err());
```
[the remark this answers is not in the document]
Comment (SSebo) [reply socketio/src/client/client.rs:468]: The panic is the old logic, I agree with you to just log.

File: socketio/src/client/builder.rs (hunk 2 above)
Comment (SSebo) [reply socketio/src/client/builder.rs:289]: @nshaaban-cPacket,  I agree it is odd. But It is hard to implement reconnecting if under laying socket creation is part of builder. user code hold the socket client, the builder is gone after `build()`, all the info about how to create socket is not available. Would you give some good suggestion?

File: socketio/src/client/client.rs (hunk 4 above)
Comment (SSebo) [reply socketio/src/client/client.rs:602]: @nshaaban-cPacket Server not start immediately, emit will not success until server is start. If I change the test server to restart immediately, user code may emit success every time, we have no idea if the server is restart or not.

File: socketio/src/client/client.rs (hunk 1 above)
Comment (nshaaban-cPacket) [reply socketio/src/client/client.rs:579]: blocker: That issue doesn't exist in main, connect should block until the client is usable.

File: socketio/src/client/builder.rs (hunk 2 above)
Comment (nshaaban-cPacket) [reply socketio/src/client/builder.rs:289]: Agreed you'll need the context of the builder in order to create a new client. Consider 
> Perhaps a wrapper around client would be cleaner separating the auto-reconnect logic from the normal client logic? Auto re-connection seems more of a concern for a wrapper around client rather than complicating the underlying client implementation.

File: socketio/src/client/client.rs (hunk 4 above)
Comment (nshaaban-cPacket) [reply socketio/src/client/client.rs:602]: (usability issue) seems difficult for users of the library to use

**Comment by SSebo** [comment]
> Like to start off by saying this is a valid implementation, as well as effort went into adding tests, which is much appreciated.

I agree with you, It is not a very clean implementation. I would like to refactor this PR.

> Concerned about the additional complication (addition of arc/mutex/rwlocks) in the underlying client, as well as increased potential in IllegalActionBeforeOpen edge cases. Neither of which are currently present on master.

Yes, but if reconnecting feature is needed, the failed socket must be replaced with new one, so there must be an arc lock. otherwise provide a notification, to let user create a new one. As you said we can try not transparent reconnect.

> Perhaps a wrapper around client would be cleaner separating the auto-reconnect logic from the normal client logic? Auto re-connection seems more of a concern for a wrapper around client rather than complicating the underlying client implementation.

Good idea.

> At the end of the day this is a valid implementation, and for how it's done, it looks good (besides is one usability question). While I may prefer a different approach, decision is up to @1c3t3a . Feel free to merge without my approval.

I would like to refactor the code, your approval is important.

> @SSebo Thanks for the PR, sorry for not documenting purpose of the connect_manual, nor intention of iterators use in the future. I recommenced waiting for @1c3t3a before attempting to implement my comments related to differing creative vision (I've annotated the comments in question), as what you have is valid and functional.
> 
> Thank you for properly implementing unit tests for your features and expanding the existing test infrastructure.

I am new here, so much to learn.

**Comment by nshaaban-cPacket** [comment]
Wonderful! I'm almost done with a PoC that should get most of the way with the wrapper I was talking about.

> Yes, but if reconnecting feature is needed, the failed socket must be replaced with new one, so there must be an arc lock. otherwise provide a notification, to let user create a new one. As you said we can try not transparent reconnect.

Yes there must be some Arc/RwLocks somewhere to get around the callbacks not being Clone, but there is some flexibility as to where ;)

**Comment by 1c3t3a** [comment]
I agree with @nshaaban-cPacket here. the proposed new solution in #229 looks better! Anyways thanks a lot for going the extra mile @SSebo on reacting to the comments here and thanks a lot @nshaaban-cPacket for your expertise and the example PR! The wrapper sounds like the perfect solution, especially since we are wrapping types in this library on each level anyways, so I think this fits the bigger picture ;)
